### PR TITLE
chips/lowrisc/spi_host: refactor and rustify.

### PR DESCRIPTION
**Pull request by twilfredo**
### Pull Request Overview

A few cleanup changes to avoid panics, reduced bloat by ~92B according to `cargo bloat`

Before:
```
 0.0%   0.5%     674B         lowrisc lowrisc::spi_host::SpiHost::handle_interrupt
```

After:
```
0.0%   0.4%     582B         lowrisc lowrisc::spi_host::SpiHost::handle_interrupt
```

Edit after a few more fixes:

```
0.0%   0.5%     632B         lowrisc lowrisc::spi_host::SpiHost::handle_interrupt
```

Could probably use a few more improvments...

### Testing Strategy

Running SPI_HOST tests on QEMU and Verilator

### TODO or Help Wanted
 
N/A

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: chips/lowrisc/src/spi_host.rs [hunk 1]
```diff
@@ -243,45 +248,78 @@ impl SpiHost {
             } else {
                 self.rx_buf.replace(rx_buf);
                 //Theres more to transfer, continue writing from the offset
-                self.spi_transfer_progress();
+                if let Err(err) = self.spi_transfer_progress() {
+                    //Transfer failed, lets clean up
+                    //Clear all pending errors.
```
Comment (hudson-ayers) [inline chips/lowrisc/src/spi_host.rs:253]:
```suggestion
                    //Clear all pending interrupts.
```

File: chips/lowrisc/src/spi_host.rs [hunk 2]
```diff
@@ -243,45 +248,78 @@ impl SpiHost {
...
+                    self.reset_internal_state();
+                    self.client.map(|client| match self.tx_buf.take() {
+                        None => (),
+                        Some(tx_buf) => client.read_write_done(
+                            tx_buf,
+                            self.rx_buf.take(),
+                            self.tx_offset.get(),
+                            Err(err),
+                        ),
+                    });
```
Comment (hudson-ayers) [inline chips/lowrisc/src/spi_host.rs:266]: we shouldn't call an upcall from within a downcall like this

**Review by hudson-ayers — CHANGES_REQUESTED** [review]
Mostly looks good but one important issue

File: chips/lowrisc/src/spi_host.rs (hunk 2 above)
Comment (twilfredo) [reply chips/lowrisc/src/spi_host.rs:266]: > we shouldn't call an upcall from within a downcall like this

Ah okay, I'll fix this up! What is the reasoning for this?

Comment (hudson-ayers) [reply chips/lowrisc/src/spi_host.rs:266]: It is justified here: https://github.com/tock/tock/blob/master/doc/reference/trd2-hil-design.md#rule-1-dont-make-synchronous-callbacks

Comment (twilfredo) [reply chips/lowrisc/src/spi_host.rs:266]: thanks!

**Comment by twilfredo** [comment]
Changes:

- Fixed up multiple instances of calling upcalls in downcalls, now they occur in the `handle_interrupt()` fn only.
- Added a new enum to detect spihost (specific to this driver) status.

**Review by alistair23 — APPROVED** [review]

**Review by hudson-ayers — APPROVED** [review]
bors r+

**Pull request merged**